swaylock/shadow.c

```diff
@@ -0,0 +1,124 @@
+#define _XOPEN_SOURCE
+#include <pwd.h>
+#include <shadow.h>
+#include <stdbool.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <wlr/util/log.h>
+#include "swaylock/swaylock.h"
+
+static int comm[2][2];
+
+void run_child() {
+	/* This code runs as root */
+	struct passwd *pwent = getpwuid(getuid());
+	if (!pwent) {
+		wlr_log_errno(WLR_ERROR, "failed to getpwuid");
+		exit(EXIT_FAILURE);
+	}
+	char *encpw = pwent->pw_passwd;
+	if (strcmp(encpw, "x") == 0) {
+		struct spwd *swent = getspnam(pwent->pw_name);
+		if (!swent) {
+			wlr_log_errno(WLR_ERROR, "failed to getspnam");
+			exit(EXIT_FAILURE);
+		}
+		encpw = swent->sp_pwdp;
+	}
+	wlr_log(WLR_DEBUG, "prepared to authorize user %s", pwent->pw_name);
+
+	size_t size;
+	char *buf;
+	while (1) {
+		if (read(comm[0][0], &size, sizeof(size)) < 0) {
```

Comment:
We could add proper error handling here. Zero means it's EOF, -1 means something bad happened.